docs/src/blockchain-development/purity.md

```diff
@@ -2,6 +2,23 @@
 
 A function is _pure_ if it does not access any [persistent storage](./storage.md). Conversely, the function is _impure_ if it does access any storage. Naturally, as storage is only available in smart contracts, impure functions cannot be used in predicates, scripts, or libraries. A pure function cannot call an impure function.
 
-In Sway, functions are pure by default but can be opted in to impurity via function annotations which are currently work-in-progress.
+In Sway, functions are pure by default but can be opted into impurity via the `storage` function attribute.  The `storage` attribute may take `read` and/or `write` arguments indicating which type of access the function requires.
+
+```sway
+#[storage(read)]
+fn get_amount() -> u64 {
+    ...
+}
+
+#[storage(read, write)]
+fn increment_amount(increment: u64) -> u64 {
+    ...
+}
+
+```
+
+Impure functions which call other impure functions must have at least the same storage privileges or a superset of those for the function called.  For example, to call a function with write access a caller must also have write access, or both read and write access.  To call a function with read and write access the caller must also have both privileges.
+
+The `storage` attribute may also be applied to [methods and associated functions](../basics/methods_and_associated_functions.md) but are not required nor accepted for [trait](../advanced/traits.md) or [ABI declarations](../sway-program-types/smart_contracts.md#the-abi-declaration).
```

Comment:
This will change in the future, right?

Reply:
Um, no?  AFAICT we decided that attributes should only be applied to implementations and not declarations.

Reply:
Looping in @sezna. That was the _initial_ decision, but it was revised (and clearly not documented 😂) to add attributes to the ABI declarations at the very least to enable method payability: https://github.com/FuelLabs/fuel-specs/issues/271. Ref: #https://github.com/FuelLabs/sway/issues/1145

> Add annotation for ABI method payability

Reply:
I'd imagine we'd want to at least support it for trait/ABI method declarations that provide default implementations, in case those default implementations need access to storage?

Otherwise my impression is that the storage access attribute is more about whether or not *implementations* of functions, trait methods or ABI methods interact with storage. It's not quite clear to me what it would mean to use the attribute with a trait or ABI method _declaration_ (that has no default implementation)?

@adlerjohn in the case of method payability, would it be enough to instead require that implementations of ABI methods that access received coins require the payable attribute, rather than requiring that the ABI declarations themselves do?

Reply:
> would it be enough to instead require that implementations of ABI methods that access received coins require the payable attribute, rather than requiring that the ABI declarations do?

No, because the _caller_ doesn't know the implementation; they only know the ABI declaration.

Reply:
> No, because the caller doesn't know the implementation; they only know the ABI declaration.

For the caller to be able to call the ABI method, there must be an implementation of the ABI for the callee somewhere, no? My understanding is that if there's an ABI declaration, for it to actually be used it must be implemented for `Contract` at some point - is that right?

If so, and if `Contract`'s ABI method implementation requires payable, then in order to call it the user would have to add the payable attribute to whatever function it is they are calling the payable method from. This would be the case whether or not the declaration has a payable attribute, right?

Reply:
> For the caller to be able to call the ABI method, there must be an implementation of the ABI for the callee somewhere, no? My understanding is that if there's an ABI declaration, for it to actually be used it must be implemented for Contract at some point - is that right?

That's correct, but just because you're correct doesn't mean you're right. The model for cross-contract calls is through an ABI, where the implementation of that ABI is unknown at compile time. This is the same model that Ethereum uses and Move (Libra's thing) does not use. It allows your contract to interact with contracts that are deployed later, where obviously it's impossible to know their implementation at compile time since they don't exist yet.

Put another way: cross-contract calls can _only_ use the ABI, not the implementation of the called contract.

Reply:
Ahhhhh, cross-contract calls was the context I was missing!

Do we have any sketches of what cross-contract calls might look like in Sway code yet?

Reply:
Yes. https://fuellabs.github.io/sway/v0.12.2/blockchain-development/calling_contracts.html

Reply:
Can I get some clarification on what still needs to be done here please?  I can add annotations to ABI declarations and update the docs now if that's what we want.

Reply:
Annotations (generally) allowed for ABI and trait method declarations. Storage annotations specifically too. And need to enforce that storage annotations for implementations match declarations.